File: swift/Sources/McpApps/Transport/Transport.swift
import Foundation

/// JSON-RPC message types for MCP Apps communication.
public enum JSONRPCMessage: Codable, Sendable {
    case request(JSONRPCRequest)
    case notification(JSONRPCNotification)
    case response(JSONRPCResponse)
    case error(JSONRPCErrorResponse)

    enum CodingKeys: String, CodingKey {
        case jsonrpc, id, method, params, result, error
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        // Check for error response first
        if container.contains(.error) {
            self = .error(try JSONRPCErrorResponse(from: decoder))
            return
        }

        // Check for result (response)
        if container.contains(.result) {
            self = .response(try JSONRPCResponse(from: decoder))
            return
        }

        // Check for id (request vs notification)
        if container.contains(.id) {
            self = .request(try JSONRPCRequest(from: decoder))
        } else {
            self = .notification(try JSONRPCNotification(from: decoder))
        }
    }

    public func encode(to encoder: Encoder) throws {
        switch self {
        case .request(let request):
            try request.encode(to: encoder)
        case .notification(let notification):
            try notification.encode(to: encoder)
        case .response(let response):
            try response.encode(to: encoder)
        case .error(let error):
            try error.encode(to: encoder)
        }
    }
}

/// JSON-RPC request message.
public struct JSONRPCRequest: Codable, Sendable {
    public var jsonrpc: String = "2.0"
    /// Unique identifier for this request
    public var id: JSONRPCId
    /// Method name to invoke
    public var method: String
    /// Optional parameters for the method
    public var params: [String: AnyCodable]?

    public init(id: JSONRPCId, method: String, params: [String: AnyCodable]? = nil) {
        self.id = id
        self.method = method
        self.params = params
    }
}

/// JSON-RPC notification message.
public struct JSONRPCNotification: Codable, Sendable {
    public var jsonrpc: String = "2.0"
    /// Method name for this notification
    public var method: String
    /// Optional parameters for the notification
    public var params: [String: AnyCodable]?

    public init(method: String, params: [String: AnyCodable]? = nil) {
        self.method = method
        self.params = params
    }
}

/// JSON-RPC success response message.
public struct JSONRPCResponse: Codable, Sendable {
    public var jsonrpc: String = "2.0"
    /// ID matching the original request
    public var id: JSONRPCId
    /// Result of the method invocation
    public var result: AnyCodable

    public init(id: JSONRPCId, result: AnyCodable) {
        self.id = id
        self.result = result
    }
}

/// JSON-RPC error response message.
public struct JSONRPCErrorResponse: Codable, Sendable {
    public var jsonrpc: String = "2.0"
    /// ID matching the original request
    public var id: JSONRPCId?
    /// Error details
    public var error: JSONRPCError

    public init(id: JSONRPCId?, error: JSONRPCError) {
        self.id = id
        self.error = error
    }
}

/// JSON-RPC error object.
public struct JSONRPCError: Codable, Sendable {
    /// Error code
    public var code: Int
    /// Human-readable error message
    public var message: String
    /// Optional additional error data
    public var data: AnyCodable?

    public init(code: Int, message: String, data: AnyCodable? = nil) {
        self.code = code
        self.message = message
        self.data = data
    }

    // Standard JSON-RPC error codes
    public static let parseError = -32700
    public static let invalidRequest = -32600
    public static let methodNotFound = -32601
    public static let invalidParams = -32602
    public static let internalError = -32603
    public static let mcpError = -32000
}

/// JSON-RPC request ID.
public enum JSONRPCId: Codable, Sendable, Hashable {
    case string(String)
    case number(Int)

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let intValue = try? container.decode(Int.self) {
            self = .number(intValue)
        } else if let stringValue = try? container.decode(String.self) {
            self = .string(stringValue)
        } else {
            throw DecodingError.typeMismatch(
                JSONRPCId.self,
                DecodingError.Context(
                    codingPath: decoder.codingPath,
                    debugDescription: "Expected string or number"
                )
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let value):
            try container.encode(value)
        case .number(let value):
            try container.encode(value)
        }
    }
}

/// Transport protocol for MCP Apps communication.
///
/// This protocol abstracts the underlying message transport mechanism,
/// allowing different implementations for various platforms.
public protocol McpAppsTransport: Actor {
    /// Start the transport and begin listening for messages.
    func start() async throws

    /// Send a JSON-RPC message to the peer.
    func send(_ message: JSONRPCMessage) async throws

    /// Close the transport and cleanup resources.
    func close() async

    /// Stream of incoming JSON-RPC messages from the peer.
    var incoming: AsyncThrowingStream<JSONRPCMessage, Error> { get }
}

/// In-memory transport for testing.
public actor InMemoryTransport: McpAppsTransport {
    private var peer: InMemoryTransport?
    private var continuation: AsyncThrowingStream<JSONRPCMessage, Error>.Continuation?

    public let incoming: AsyncThrowingStream<JSONRPCMessage, Error>

    private init(peer: InMemoryTransport?) {
        var continuation: AsyncThrowingStream<JSONRPCMessage, Error>.Continuation?
        self.incoming = AsyncThrowingStream { continuation = $0 }
        self.continuation = continuation
        self.peer = peer
    }

    public func start() async throws {
        // Nothing to do for in-memory transport
    }

    public func send(_ message: JSONRPCMessage) async throws {
        guard let peer = peer else {
            throw TransportError.notConnected
        }
        await peer.receiveFromPeer(message)
    }

    public func close() async {
        continuation?.finish()
        peer = nil
    }

    func receiveFromPeer(_ message: JSONRPCMessage) {
        continuation?.yield(message)
    }

    func setPeer(_ peer: InMemoryTransport) {
        self.peer = peer
    }

    /// Create a linked pair of transports for testing.
    public static func createLinkedPair() async -> (InMemoryTransport, InMemoryTransport) {
        let first = InMemoryTransport(peer: nil)
        let second = InMemoryTransport(peer: first)
        await first.setPeer(second)
        return (first, second)
    }
}

/// Transport errors.
public enum TransportError: Error {
    case notConnected
    case serializationFailed
    case invalidMessage
}

File: kotlin/src/main/kotlin/io/modelcontextprotocol/apps/transport/Transport.kt
package io.modelcontextprotocol.apps.transport

import io.modelcontextprotocol.apps.protocol.JSONRPCMessage
import kotlinx.coroutines.flow.Flow

/**
 * Transport interface for MCP Apps communication.
 *
 * This interface abstracts the underlying message transport mechanism,
 * allowing different implementations for various platforms:
 * - WebView (Android/iOS) using JavaScript bridges
 * - In-memory for testing
 * - postMessage for web (via TypeScript SDK)
 */
interface McpAppsTransport {
    /**
     * Start the transport and begin listening for messages.
     *
     * This should set up any necessary listeners or bridges.
     */
    suspend fun start()

    /**
     * Send a JSON-RPC message to the peer.
     *
     * @param message The JSON-RPC message to send
     */
    suspend fun send(message: JSONRPCMessage)

    /**
     * Close the transport and cleanup resources.
     */
    suspend fun close()

    /**
     * Flow of incoming JSON-RPC messages from the peer.
     */
    val incoming: Flow<JSONRPCMessage>

    /**
     * Flow of transport errors.
     */
    val errors: Flow<Throwable>
}

/**
 * In-memory transport for testing.
 *
 * Creates a pair of linked transports that forward messages to each other.
 */
class InMemoryTransport private constructor(
    private val peer: InMemoryTransport?
) : McpAppsTransport {

    private val _incoming = kotlinx.coroutines.flow.MutableSharedFlow<JSONRPCMessage>()
    private val _errors = kotlinx.coroutines.flow.MutableSharedFlow<Throwable>()

    private var _peer: InMemoryTransport? = peer

    override val incoming: Flow<JSONRPCMessage> = _incoming
    override val errors: Flow<Throwable> = _errors

    override suspend fun start() {
        // Nothing to do for in-memory transport
    }

    override suspend fun send(message: JSONRPCMessage) {
        _peer?._incoming?.emit(message)
            ?: throw IllegalStateException("Transport not connected to peer")
    }

    override suspend fun close() {
        _peer = null
    }

    companion object {
        /**
         * Create a linked pair of transports for testing.
         *
         * Messages sent on one transport are received on the other.
         *
         * @return A pair of linked transports (first, second)
         */
        fun createLinkedPair(): Pair<InMemoryTransport, InMemoryTransport> {
            val first = InMemoryTransport(null)
            val second = InMemoryTransport(first)
            first._peer = second
            return first to second
        }
    }
}
